Add setting options data store for managing settings

chihsuan opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

This PR introduces a new `settings-options` data store that provides a centralized way to manage WooCommerce settings. The store handles settings groups, individual settings, unsaved changes, and error states.

#### Key features

- Manage settings groups and individual settings
- Handle unsaved changes with edit states
- Batch update capabilities
- Properly handle errors for failed updates

#### Why this is needed?

We've deprecated the `option` data store in favor of creating individual endpoints for each usage. However, having a new endpoint for each setting is ineffective. I was trying to [migrate the core profiler to use dedicated endpoints for saving store settings](https://github.com/woocommerce/woocommerce/pull/56347) but realized that we can reuse the existing [WC `setting options` API](https://woocommerce.github.io/woocommerce-rest-api-docs/#setting-options).

With this new data store, we can easily manage settings groups and individual settings. It's also useful for new modern Settings pages except for replacing the existing Settings/Options store.

#### Why introduce a new data store rather than use the existing `settings` store?

The existing `settings` store was introduced for the same purpose. However, it has many issues:

1. The method signatures are confusing. For example, `select( settingsStore ).getSettings('general').general;` is not very intuitive and `wp.data.select( wc.data.settingsStore ).getSetting( 'general', 'woocommerce_default_country' )`
2. It's mixed with other unrelated settings because we preload all wcSettings.admin to the store, so you can use `resolveSelect( settingsStore ).getSettings( 'wc_admin').shopUrl` to get the store URL, but you can't update it since it's setting in WC > Settings.
3. It doesn't handle errors correctly. For example, when updating settings in a batch and some fail, it doesn't return the error. Instead, it considers API success.

Besides, I [tried to enhance the existing `settings` store](https://github.com/woocommerce/woocommerce/pull/56478), but it's pretty complicated since we also need to ensure backward compatibility. I believe it's easier to create a new data store for this and deprecate the existing one in the future.


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

**This PR doesn't add/update any new user-facing features. I'll have some follow-up PRs to migrate the code to use this new data store.**

Review the [REAMDE.md](https://github.com/woocommerce/woocommerce/blob/add/settings-option-store/packages/js/data/src/setting-options/README.md)

1. Create a fresh site 
2. Go to any WooCommerce Admin page
3. Open dev console and test the following commands:

#### Basic Operations

```js
wp.data.resolveSelect( wc.data.experimentalSettingOptionsStore ).getGroups().then(console.log)
// should print an array of groups

wp.data.resolveSelect( wc.data.experimentalSettingOptionsStore ).getGroup( 'general' ).then(console.log)
// should print an object of general group

wp.data.resolveSelect( wc.data.experimentalSettingOptionsStore ).getSettings( 'general' ).then(console.log)
// should print an object of general settings (e.g. woocommerce_default_country...)
```

#### Single Setting Operations

```js
wp.data.resolveSelect( wc.data.experimentalSettingOptionsStore ).getSetting( 'general', 'woocommerce_currency' ).then(console.log)
// Should print the woocommerce_currency setting object

wp.data.resolveSelect( wc.data.experimentalSettingOptionsStore ).getSettingValue( 'general', 'woocommerce_currency' ).then(console.log)
// Should print the woocommerce_currency setting value
```

#### Batch Operations

```js
// Update multiple settings at once
wp.data.dispatch( wc.data.experimentalSettingOptionsStore ).editSettings( 'general', { woocommerce_currency: 'JPY', 'woocommerce_default_country': 'GB'  } )

wp.data.select( wc.data.experimentalSettingOptionsStore ).getEditedSettingIds( 'general')
// Should return  ['woocommerce_currency', 'woocommerce_default_country']

wp.data.dispatch( wc.data.experimentalSettingOptionsStore ).saveEditedSettingsGroup( 'general' );
// Confirm it send the API request to update the settings and return 200 status in network tab
```

#### Direct Save Operations

```js
wp.data.dispatch( wc.data.experimentalSettingOptionsStore ).saveSettingsGroup( 'general', { woocommerce_currency: 'USD', 'woocommerce_default_country': 'US:CA'  } )
// Confirm it send the API request to update the settings and return 200 status in network tab

wp.data.dispatch( wc.data.experimentalSettingOptionsStore ).saveSetting( 'general', 'woocommerce_currency', 'JPY' )
// Confirm it send the API request to update the settings and return 200 status in network tab
```

#### Error Handling

```js
wp.data.dispatch( wc.data.experimentalSettingOptionsStore ).saveSetting( 'general', 'woocommerce_currency', '' )
// Confirm it throw an error in console

wp.data.select( wc.data.experimentalSettingOptionsStore ).getGroupError( 'general' )
// Confirm it return the object with key `woocommerce_currency`

wp.data.select( wc.data.experimentalSettingOptionsStore ).getSettingError( 'general', 'woocommerce_currency')
// Confirm it return the error object
```

#### Testing Concurrent Operations

```js
// Refresh the page to reset the state
// Test 1: Concurrent save and fetch
Promise.all([
  wp.data.dispatch( wc.data.experimentalSettingOptionsStore ).saveSetting( 'general', 'woocommerce_currency', 'JPY' ),
  wp.data.resolveSelect( wc.data.experimentalSettingOptionsStore ).getSetting( 'general', 'woocommerce_currency' )]
).then(console.log);
// Check network tab to observe the request order (screenshot below) and final state

// Test 2: Multiple concurrent saves
Promise.all([ 
  wp.data.dispatch( wc.data.experimentalSettingOptionsStore ).saveSetting( 'general', 'woocommerce_currency', 'JPY' ),
  wp.data.dispatch( wc.data.experimentalSettingOptionsStore ).saveSetting( 'general', 'woocommerce_currency', 'USD' )
]).then(console.log);
// Verify final currency value and check network requests

// Test 3: Edit during save
wp.data.dispatch( wc.data.experimentalSettingOptionsStore ).saveSetting( 'general', 'woocommerce_currency', 'JPY' );
wp.data.dispatch( wc.data.experimentalSettingOptionsStore ).editSetting( 'general', 'woocommerce_currency', 'USD' );
// Check if edit is preserved or overwritten by save response
wp.data.select( wc.data.experimentalSettingOptionsStore ).getEditedSettingIds( 'general' )
// Should return ['woocommerce_currency']
```

❌
![Screenshot 2025-03-26 at 14 35 33](https://github.com/user-attachments/assets/992367d5-d663-4950-a87e-a84e29f10c55)

✅ 
![Screenshot 2025-03-26 at 14 36 45](https://github.com/user-attachments/assets/009af55a-df09-4213-a8fa-015e07828dc8)


### Expected Results for Concurrent Operations

1. When running save and fetch concurrently, the fetch should return the most up-to-date value
2. When running multiple saves concurrently, the last save operation should be reflected in the final state
3. When editing during a save operation, the edit should not be lost due to the save response

### Things to Watch For:

- Check Network tab to observe request timing and order
- Monitor console for any errors or race conditions
- Verify final state matches expected values

Please try different combinations of dispatching actions and select the data you want to test.


<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

chihsuan committed:
Rename saveSetting and saveSettingsGroup actions to saveEditedSetting and saveEditedSettingsGroup for clarity and consistency.

chihsuan committed:
Rename settings options data store to experimental setting options and update README for clarity

chihsuan committed:
Enhance getSettings selector to apply edits to settings

- Updated the `getSettings` selector to return settings with applied edits when they exist.
- Modified the logic to create a new object that merges original settings with edited values.
- Added tests to verify that settings reflect edits correctly in various scenarios.

chihsuan committed:
Refactor action types and selectors for settings options

- Updated action type definitions to use `CurriedSelectorsOf` for improved type safety and consistency.
- Removed unused `DispatchFromMap` type and streamlined the `ThunkArgs` type structure.
- Enhanced code clarity by simplifying type definitions related to selectors.

chihsuan committed:
Add settings options data store for managing WooCommerce settings

- Introduced a new data store for settings options, including actions, selectors, and reducers.
- Implemented action types for receiving and updating settings and groups.
- Added resolvers for fetching settings and groups from the API.
- Created selectors for accessing settings and their states.
- Included tests for actions, reducers, and selectors to ensure functionality.
- Updated index files to export the new store and its components.
- Added documentation for usage and API interactions.

chihsuan committed:
Add changelog

chihsuan committed:
Fix action

chihsuan committed:
Update readme

chihsuan committed:
Fix tests

chihsuan committed:
Fix lint

chihsuan committed:
Remove fallback error handling in getGroups resolver

chihsuan committed:
Refactor selectors for settings options to use memoization and ensure consistent return values

- Introduced `createSelector` from `rememo` for `getSetting` and `getEditedSettingIds` to optimize performance.
- Replaced empty object return in `getSettings` with a consistent `EMPTY_OBJECT`.
- Improved code readability and maintainability by restructuring selector logic.

rjchow commented on packages/js/data/src/setting-options/actions.ts:
```diff
@@ -0,0 +1,361 @@
+ * External dependencies
+ */
+import apiFetch from '@wordpress/api-fetch';
+import { DispatchFromMap } from '@automattic/data-stores';
```
I suspect the types in `@automattic/data-stores` are from before `@wordpress/data` started exporting its own types, so using the types from there will result in conflicts at a later time - do you know if that's the case?

chihsuan replied:
We only export ActionDispatchers without overriding @wordpress/data types, so I believe using the types from there is safe. But actually, I think we don't need this. It was needed for the old way of doing things. I've removed it!

rjchow commented on packages/js/data/src/setting-options/actions.ts:
```diff
@@ -0,0 +1,361 @@
+
+type WPDataRegistry = ReturnType< typeof createRegistry >;
+type AllSelectors = typeof import('./selectors');
+type CurriedState< F > = F extends (
```
I thiiiiiiiink you can get this from `@wordpress/data`

chihsuan replied:
Thanks! Updated in 4711d41d253a6383ce37626d6423f75e1cc7bbc2.

rjchow commented on packages/js/data/src/setting-options/actions.ts:
```diff
@@ -0,0 +1,361 @@
+) => infer R
+	? ( ...args: P ) => R
+	: F;
+type Selectors = {
```
This too

rjchow commented on packages/js/data/src/setting-options/selectors.ts:
```diff
@@ -0,0 +1,213 @@
+): SettingsGroup | undefined =>
+	state.groups.find( ( group ) => group.id === groupId );
+
+// Ensure we have a consistent empty object to return when there are no settings.
```
Is this for checking for referential equality?

chihsuan replied:
Yes, selectors need to return a scalar or a consistent reference. https://developer.wordpress.org/news/2024/03/how-to-work-effectively-with-the-useselect-hook

chihsuan committed:
Enhance settings options reducer and types for improved state management

- Added a utility function `ensureGroupExists` to streamline group existence checks in the reducer.
- Updated the reducer to utilize the new utility for managing settings, edits, and saving states.
- Modified types to allow for optional group and setting structures, improving type safety.
- Enhanced error handling and state tracking in the reducer.
- Updated tests to reflect changes in state structure and ensure proper functionality.

rjchow commented on packages/js/data/src/setting-options/tests/actions.test.ts:
```diff
@@ -0,0 +1,615 @@
+				mockResults.update[ 0 ]
+			);
+
+			// Verify error state
```
In conjunction with the other comment about not caching error values, if that's accepted then we should also verify that the returned value is correct here

rjchow reviewed (commented):
- after an error in saving, e.g saving woocommerce_currency to '', the getSetting and getSettingsValue returns the invalid value - I think the cache should not be saved if the saving fails?
- should updateSettings be stageSettingUpdate? I think its slightly unintuitive that an imperative verb like 'update' doesn't persist the settings by default, and requires a parameter to make it save
- when doing a refresh after an error occurs, this weird thing shows up in the console not really sure if its an issue

![image](https://github.com/user-attachments/assets/2116c83a-0350-4caf-bf2c-fb30593aa617)

Otherwise, really nice job! I think the API is quite comprehensively thought out and the test suite is very good. I wonder if we should mark this experimental somehow while we iterate on it? I reckon we might learn that we'll need some changes when we work on the modern Settings work

rjchow commented on packages/js/data/src/setting-options/types.ts:
```diff
@@ -0,0 +1,160 @@
+/**
+ * Represents the value of a setting. It's unknown because any value can be stored.
+ */
+export type SettingValue = unknown;
```
Gut feel that having it as unknown might result in a bunch of type overrides but I'm happy to leave it alone for now and see if it proves true

chihsuan replied:
It appears that the majority of settings values are strings, although `wc_admin` utilizes `string[]` for a few. Additionally, it seems that any type of value can be registered so I left it as unknown for now.

chihsuan commented:
Thanks for the review! @rjchow 🙏 

> * after an error in saving, e.g saving woocommerce_currency to '', the getSetting and getSettingsValue returns the invalid value - I think the cache should not be saved if the saving fails?

If a user changes a setting and the save operation fails, I think the edited state should still be cached instead of clearing it so that the user can see the edited value with the error message. Does this make sense? I think it's common for users to edit an input field, and then the save operation fails, but they still see the edited value.

> * should updateSettings be stageSettingUpdate? I think its slightly unintuitive that an imperative verb like 'update' doesn't persist the settings by default, and requires a parameter to make it save

Sounds good!

> * when doing a refresh after an error occurs, this weird thing shows up in the console not really sure if its an issue
> 
> ![image](https://private-user-images.githubusercontent.com/27843274/424974314-2116c83a-0350-4caf-bf2c-fb30593aa617.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDI1Mzg2MDgsIm5iZiI6MTc0MjUzODMwOCwicGF0aCI6Ii8yNzg0MzI3NC80MjQ5NzQzMTQtMjExNmM4M2EtMDM1MC00Y2FmLWJmMmMtZmIzMDU5M2FhNjE3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzIxVDA2MjUwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMyMTM4YTIyNzAwODgyMzkyZjc1Yzk2NTc0NmFmOWEzNDE1OTAyNmFlYTdlMjAyYjcyNzMxOTA3NTViYjUwZjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.y5j2rN4Dh6B-9r1CburxHxLczn_hQD_hBSenmMMEYf4)
> 

I'll test more and see if I can reproduce it. I remember that if you have redux devtools open, it will show some weird things because of the extension cache.

chihsuan commented:
> I wonder if we should mark this experimental somehow while we iterate on it? I reckon we might learn that we'll need some changes when we work on the modern Settings work


Good idea, I'll export the store as experimental and add a note to the readme.

rjchow commented:
> > * after an error in saving, e.g saving woocommerce_currency to '', the getSetting and getSettingsValue returns the invalid value - I think the cache should not be saved if the saving fails?
> 
> If a user changes a setting and the save operation fails, I think the edited state should still be cached instead of clearing it so that the user can see the edited value with the error message. Does this make sense? I think it's common for users to edit an input field, and then the save operation fails, but they still see the edited value.

I think this use case makes sense - do we need to handle the situation where multiple instances of select/dispatch on the same store is being used at the same time?


> > * should updateSettings be stageSettingUpdate? I think its slightly unintuitive that an imperative verb like 'update' doesn't persist the settings by default, and requires a parameter to make it save
> 
> Sounds good! What do you think if we rename `updateSettings` to `editSettings`? WP core data has similar pattern, e.g. `saveEntityRecord` and `editEntityRecord`. I'm not very sure if this is a good idea from a natural language perspective. 🤔
 
I think that's fine!

chihsuan committed:
Rename updateSetting* -> editSetting*

chihsuan commented:
> > If a user changes a setting and the save operation fails, I think the edited state should still be cached instead of clearing it so that the user can see the edited value with the error message. Does this make sense? I think it's common for users to edit an input field, and then the save operation fails, but they still see the edited value.
> 
> I think this use case makes sense - do we need to handle the situation where multiple instances of select/dispatch on the same store is being used at the same time?

Hmm, good point. I'll need to think about this.

> > Sounds good! What do you think if we rename `updateSettings` to `editSettings`? WP core data has similar pattern, e.g. `saveEntityRecord` and `editEntityRecord`. I'm not very sure if this is a good idea from a natural language perspective. 🤔
> 
> I think that's fine!

Ok, I've renamed `updateSetting*` to `editSetting*`.

psealock commented on packages/js/data/src/setting-options/actions.ts:
```diff
@@ -0,0 +1,345 @@
+					result.error &&
+					typeof result.error === 'object' &&
+					'code' in result.error
+				) {
```
Is every error guaranteed to always have a `code` property? I'm wondering if this check is overly specific and may miss an error.

chihsuan replied:
I confirmed that the WC API ensures every error includes a code property. But your concern is valid. The check might be too stringent. I've removed it in 323a842f9659051a0e997db3c10cfa1e6fb97c79.

psealock reviewed (approved):
This is really nice @chihsuan, thanks for making it happen.

I tested on JN instance by calling selectors and dispatches via the console. Just a small question, but no real issues pop out - this is testing well.

chihsuan committed:
Simplify error checking logic

chihsuan committed:
Refactor setting options actions and types for clarity

- Renamed action types from `UPDATE_SETTING` and `UPDATE_SETTINGS` to `EDIT_SETTING` and `EDIT_SETTINGS` for consistency.
- Updated action creators and reducer logic to reflect the new naming conventions.
- Enhanced documentation in README and tests to align with the changes.
- Introduced new actions for saving settings directly without requiring edits first.
- Improved error handling and state management for edited settings.

chihsuan committed:
Enhance setting options selectors to include edits functionality

- Updated `getSettings`, `getSetting`, and `getSettingValue` selectors to accept an `options` parameter for including edits.
- Modified the logic to return original settings or edited values based on the `includeEdits` option.
- Updated README documentation to reflect changes in selector usage.
- Added tests to verify the correct behavior of selectors with and without edits.

chihsuan committed:
Fix lint

chihsuan committed:
Refactor setting options actions and reducer for improved state management

- Introduced action creators for editing single and multiple settings, enhancing clarity and reusability.
- Updated the reducer to handle edits more efficiently, ensuring edits are only stored when values change and removing them when they revert to original values.
- Added utility functions to manage empty groups in the state, improving state cleanliness.
- Enhanced tests to cover new functionality and ensure correct behavior of edits and error handling.

chihsuan committed:
Implement store locking mechanism for setting options actions

- Added store locking functionality to `saveSettingRequest` and `saveSettingsGroupRequest` actions to prevent concurrent modifications.
- Integrated locking in `getSettings` and `getSetting` resolvers to ensure data consistency during fetch operations.
- Updated action dispatchers to include lock actions from core data for better state management.
- Enhanced type definitions to reflect the new locking actions.

chihsuan committed:
Fix tests

chihsuan commented:
Hey @rjchow, I've made some changes: 

1.  Modified the logic to return original settings or edited values based on the `includeEdits` option so developers can decide whether to get edited values.
2. Added new actions to save settings directly instead of using the { save } option in edit* actions.
3. Integrated lock actions from core data to prevent concurrent operations on the same setting. I imported the `createLocksActions` directly from core data build folder. Alternatively, we can copy the code from Core Data to our package.
4. Some naming changes to make it more consistent and easier to understand.

Could you review the changes again? Let me know if you have any questions or feedback. 🙏 

Thanks!

> when doing a refresh after an error occurs, this weird thing shows up in the console not really sure if its an issue

For this, I couldn't reproduce it. I don't think we cache any data in the store so every time we refresh, it should be like a fresh state. 🤔

chihsuan committed:
Improve types

chihsuan committed:
Fix README.md

rjchow reviewed (commented):
I think there's some issues with the cache, but otherwise it looks good and almost ready to merge
```
wp.data.dispatch( wc.data.experimentalSettingOptionsStore ).editSettings( 'general', { woocommerce_currency: 'AED', 'woocommerce_default_country': 'GB'  } )
Promise { <state>: "pending" }

wp.data.dispatch( wc.data.experimentalSettingOptionsStore ).saveEditedSettingsGroup( 'general' );
Promise { <state>: "pending" }

wp.data.resolveSelect( wc.data.experimentalSettingOptionsStore ).getSetting( 'general', 'woocommerce_currency' ).then(console.log)
Object { id: "woocommerce_currency", label: "Currency", description: "This controls what currency prices are listed at in the catalog and which currency gateways will take payments in.", type: "select", default: "USD", options: {…}, tip: "This controls what currency prices are listed at in the catalog and which currency gateways will take payments in.", value: "JPY", group_id: "general", _links: {…} }

Promise { <state>: "fulfilled", <value>: undefined }

wp.data.resolveSelect( wc.data.experimentalSettingOptionsStore ).getSettingValue( 'general', 'woocommerce_currency' ).then(console.log)
Promise { <state>: "pending" }

AED
wp.data.resolveSelect( wc.data.experimentalSettingOptionsStore ).getSetting( 'general', 'woocommerce_currency' ).then(console.log)
Object { id: "woocommerce_currency", label: "Currency", description: "This controls what currency prices are listed at in the catalog and which currency gateways will take payments in.", type: "select", default: "USD", options: {…}, tip: "This controls what currency prices are listed at in the catalog and which currency gateways will take payments in.", value: "JPY", group_id: "general", _links: {…} }

Promise { <state>: "fulfilled", <value>: undefined }
```

chihsuan committed:
Fix setting options selectors and add tests

- Updated `getSettings` and `getSetting` selectors to default `options` to include edits functionality.
- Modified logic to return edited values based on the `includeEdits` option.
- Added unit tests to verify memoization and behavior changes when settings and `includeEdits` option are modified.

chihsuan committed:
Improve wpModulesMapper to exclude mappings for paths with suffixes

chihsuan commented:
> I think there's some issues with the cache, but otherwise it looks good and almost ready to merge
> 
> ```
> wp.data.dispatch( wc.data.experimentalSettingOptionsStore ).editSettings( 'general', { woocommerce_currency: 'AED', 'woocommerce_default_country': 'GB'  } )
> Promise { <state>: "pending" }
> 
> wp.data.dispatch( wc.data.experimentalSettingOptionsStore ).saveEditedSettingsGroup( 'general' );
> Promise { <state>: "pending" }
> 
> wp.data.resolveSelect( wc.data.experimentalSettingOptionsStore ).getSetting( 'general', 'woocommerce_currency' ).then(console.log)
> Object { id: "woocommerce_currency", label: "Currency", description: "This controls what currency prices are listed at in the catalog and which currency gateways will take payments in.", type: "select", default: "USD", options: {…}, tip: "This controls what currency prices are listed at in the catalog and which currency gateways will take payments in.", value: "JPY", group_id: "general", _links: {…} }
> 
> Promise { <state>: "fulfilled", <value>: undefined }
> 
> wp.data.resolveSelect( wc.data.experimentalSettingOptionsStore ).getSettingValue( 'general', 'woocommerce_currency' ).then(console.log)
> Promise { <state>: "pending" }
> 
> AED
> wp.data.resolveSelect( wc.data.experimentalSettingOptionsStore ).getSetting( 'general', 'woocommerce_currency' ).then(console.log)
> Object { id: "woocommerce_currency", label: "Currency", description: "This controls what currency prices are listed at in the catalog and which currency gateways will take payments in.", type: "select", default: "USD", options: {…}, tip: "This controls what currency prices are listed at in the catalog and which currency gateways will take payments in.", value: "JPY", group_id: "general", _links: {…} }
> 
> Promise { <state>: "fulfilled", <value>: undefined }
> ```

Good catch! @rjchow Confirmed that the issue is related to the selector cache not being correctly invalidated. I've fixed the selectors' memoization and added some tests for it in https://github.com/woocommerce/woocommerce/pull/56516/commits/63d8d3851b116272bb6aaf5be43fabb1389e6a77.

![Screenshot 2025-03-27 at 14 35 12](https://github.com/user-attachments/assets/db986643-e81d-4eba-ab9d-ca1b737dd8d3)

rjchow reviewed (approved):
Thanks! Magnificent work!
